README.md
SmushIt
==========

[![Version](https://img.shields.io/packagist/v/davgothic/smushit.svg)](https://packagist.org/packages/davgothic/smushit)
[![Build Status](https://img.shields.io/travis/davgothic/smushit.svg)](https://travis-ci.org/davgothic/smushit)
[![Downloads](https://img.shields.io/packagist/dt/davgothic/smushit.svg)](https://packagist.org/packages/davgothic/smushit)
[![License](https://img.shields.io/github/license/davgothic/smushit.svg)](https://github.com/davgothic/smushit/blob/master/LICENSE.md)

SmushIt is a PHP client for the popular Yahoo! image compression web service [Smush.it](http://www.smushit.com/ysmush.it/)

Basic Usage
-----------

```php
use DavGothic\SmushIt\Client;
use DavGothic\SmushIt\SmushIt;

include __DIR__ . '/vendor/autoload.php';

$client = new Client\Curl();
$smushit = new SmushIt($client);

// Compress a local/remote image and return the result object.
$result = $smushit->compress('some/path/to/an/image.png');
print_r($result);

// stdClass Object
// (
//     [src] => http://static0.resmush.it/output/1262dc777d8b239cfdf5f528a4032f02/source.png
//     [dest] => http://static1.resmush.it/output/a9ba82e7ba18e9482e085fadb126edad/output.png
//     [src_size] => 455200
//     [dest_size] => 158075
//     [percent] => 65
//     [format] => png
//     [expires] => Sun, 19 Mar 2017 18:00:33 +0100
//     [generator] => reSmush.it rev.1.4.22.20170224
// )
```

Installation
------------

To install the most recent version via [composer](https://getcomposer.org/), run the following command:

```sh
composer require davgothic/smushit
```

Requirements
------------

 - PHP 5.5.0+
 - PHP JSON extension
 - PHP cURL extension (This requirement can be ignored if using any client other than the provided cURL client)


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [3.0.1] - 2018-08-05
### Changed
- Tidied up code style
- Added ext-json to composer.json requirements in place of extension_loaded check

## [3.0.0] - 2018-02-10
### Added
- Added client interface to allow custom client implementations
- Added Guzzle HTTP client implementation
- Added unit tests
- Added Travis-CI integration

### Changed
- Moved cURL client implementation out of the SmushIt class
- Moved cURL PHP extension check out of SmushIt class and into cURL client

## [2.0.1] - 2018-01-30
### Added
- Added CHANGELOG.md
- Added installation section to README.md

### Removed
- Removed change log entries from README.md
- Removed License entry from README.md

### Fixed
- Added missing DocBlock param for SmushItException constructor

## [2.0.0] - 2017-03-19
### Changed
- Updated coding standards to follow PSR1/2
- Refactored to make use of PSR4 autoloading
- Increased required PHP version to 5.5.0
- Switched to semantic versioning of releases

## [1.3] - 2016-08-10
### Changed
- Updated webservice URI

## [1.2] - 2015-06-17
### Changed
- Updated webservice URI

## [1.1] - 2011-04-15
### Added
- Added request throttling (Thanks [Elan Ruusamäe](https://github.com/glensc))

## [1.0] - 2011-04-14
### Added
- Initial public release


tests/SmushItTest.php
<?php

use DavGothic\SmushIt\Exception\SmushItException;
use DavGothic\SmushIt\SmushIt;

/**
 * Corresponding Class to test SmushIt class
 *
 * @author David Hancock <davgothic@gmail.com>
 */
class SmushItTest extends \PHPUnit\Framework\TestCase
{

    /**
     * @var SmushIt The SmushIt object.
     */
    private $smushIt;

    public function setUp()
    {
        $responseArray = [
            'src'       => 'http://static0.resmush.it/output/1262dc777d8b239cfdf5f528a4032f02/source.png',
            'dest'      => 'http://static1.resmush.it/output/a9ba82e7ba18e9482e085fadb126edad/output.png',
            'src_size'  => 455200,
            'dest_size' => 158075,
            'percent'   => 65,
            'format'    => 'png',
            'expires'   => 'Sun, 19 Mar 2017 18:00:33 +0100',
            'generator' => 'reSmush.it rev.1.4.22.20170224',
        ];

        $client = $this->getMock(\DavGothic\SmushIt\Client\ClientInterface::class);

        $client->expects($this->any())
               ->method('execute')
               ->willReturn(json_encode($responseArray));

        /** @var \DavGothic\SmushIt\Client\Client $client */
        $this->smushIt = new SmushIt($client);
    }

    public function testIfRemoteImageSmushed()
    {
        $image  = 'https://example.org/image.jpg';
        $result = $this->smushIt->compress($image);
        $this->assertTrue(isset($result->dest));
    }

    public function testIfLocalImageSmushed()
    {
        $image  = 'tests/data/test.png';
        $result = $this->smushIt->compress($image);
        $this->assertTrue(isset($result->dest));
    }

    public function testIfExceptionThrownForInvalidImagePath()
    {
        $this->setExpectedException(SmushItException::class);
        $this->smushIt->compress('');
    }

    public function testIfExceptionThrownOnEmptyResponse()
    {
        $this->setExpectedException(SmushItException::class);
        $this->invokeMethod($this->smushIt, 'parseResponse', [null]);
    }

    public function testIfExceptionThrownOnErrorResponse()
    {
        $this->setExpectedException(SmushItException::class);
        $this->invokeMethod($this->smushIt, 'parseResponse', ['{"error":400,"error_long":"message"}']);
    }

    /**
     * Call protected/private method of a class.
     *
     * @param object $object Instantiated object that we will run method on.
     * @param string $methodName Method name to call
     * @param array $parameters Array of parameters to pass into method.
     *
     * @return mixed Method return.
     */
    public function invokeMethod(&$object, $methodName, array $parameters = [])
    {
        $reflection = new \ReflectionClass(get_class($object));
        $method     = $reflection->getMethod($methodName);
        $method->setAccessible(true);

        return $method->invokeArgs($object, $parameters);
    }

}


tests/Client/ClientTest.php
<?php

use DavGothic\SmushIt\Client\Client;

/**
 * Corresponding Class to test SmushIt class
 *
 * @author David Hancock <davgothic@gmail.com>
 */
class ClientTest extends \PHPUnit\Framework\TestCase
{

    /**
     * @var Client The Client object.
     */
    private $client;

    public function setUp()
    {
        $this->client = $this->getMockForAbstractClass(Client::class);
    }

    public function testSettingTimeout()
    {
        $timeout = 10;
        $this->client->setTimeout($timeout);

        $this->assertEquals($timeout, $this->client->getTimeout());
    }

    public function testSettingUserAgent()
    {
        $userAgent = 'Test User Agent';
        $this->client->setUserAgent($userAgent);

        $this->assertEquals($userAgent, $this->client->getUserAgent());
    }

}


tests/Exception/SmushItExceptionTest.php
<?php

use DavGothic\SmushIt\Exception\SmushItException;

/**
 * Corresponding Class to test SmushItException class
 *
 * @author David Hancock <davgothic@gmail.com>
 */
class SmushItExceptionTest extends \PHPUnit\Framework\TestCase
{

    public function testExceptionImagePathDoesNotChange()
    {
        $imagePath = '/test/image/path.png';
        $exception = new SmushItException('', 0, $imagePath);
        $this->assertSame($imagePath, $exception->getImage());
    }

}


src/SmushIt.php
<?php

namespace DavGothic\SmushIt;

use DavGothic\SmushIt\Client\Client;
use DavGothic\SmushIt\Client\ClientInterface;
use DavGothic\SmushIt\Exception\SmushItException;

/**
 * SmushIt - A PHP client for the Yahoo! Smush.it web service.
 *
 * @version    3.0.1
 * @author     David Hancock <davgothic@gmail.com>
 * @author     Elan Ruusamäe <glen@delfi.ee>
 * @copyright  (c) 2011 David Hancock
 * @license    MIT
 * @link       http://github.com/davgothic/SmushIt
 */
class SmushIt
{

    // URL to the Smush.it web service.
    const SMUSH_URL = 'http://api.resmush.it/ws.php?';

    // User agent string to set for the request.
    const USER_AGENT = 'ShushIt PHP Client/3.0.1 (+http://github.com/davgothic/SmushIt)';
    /**
     * @var int How often it is allowed to send requests. In microseconds.
     */
    public $requestInterval = 1000000;
    /**
     * @var string location of the image.
     */
    private $imageLocation;
    /**
     * @var ClientInterface The client object.
     */
    private $client;
    /**
     * @var int Time of last request.
     */
    private $requestTime;

    /**
     * Create a new SmushIt instance.
     *
     * @param ClientInterface $client The client to use for the request.
     *
     * @throws \RuntimeException
     */
    public function __construct(ClientInterface $client)
    {
        $this->client = $client;
    }

    /**
     * @param string $imageLocation The location of the image.
     *
     * @throws \DavGothic\SmushIt\Exception\SmushItException
     * @return \stdClass
     *
     *  src       = Source location of the input image.
     *  src_size  = Size of the source image in bytes.
     *  dest      = Temporary location of the compressed image.
     *  dest_size = Size of the compressed image in bytes.
     *  percent   = How much smaller the compressed image is.
     *  expires   = The date when the file will be deleted from the server
     */
    public function compress($imageLocation)
    {
        // Check if we should throttle the request to once per $requestInterval
        // @codeCoverageIgnoreStart
        if (!empty($this->requestTime)) {
            $sinceLast = ((microtime(true) - $this->requestTime) * 1000000);

            if ($sinceLast < $this->requestInterval) {
                // Sleep it off
                usleep($this->requestInterval - $sinceLast);
            }
        }
        // @codeCoverageIgnoreEnd

        $this->requestTime = microtime(true);

        $this->imageLocation = $imageLocation;

        if (null === $this->client->getUserAgent()) {
            $this->client->setUserAgent(self::USER_AGENT);
        }

        if (preg_match('/https?:\/\//', $this->imageLocation) === 1) {
            $result = $this->smushUrl();
        } else {
            $result = $this->smushFile();
        }

        return $result;
    }

    /**
     * Compress a remote image using the Smush.it web service.
     *
     * @throws \DavGothic\SmushIt\Exception\SmushItException
     * @return \stdClass
     */
    private function smushUrl()
    {
        $jsonStr = $this->client->execute(Client::TYPE_REMOTE, $this->imageLocation);

        return $this->parseResponse($jsonStr);
    }

    /**
     * Compress a local image using the Smush.it web service.
     *
     * @throws SmushItException
     * @return \stdClass
     */
    private function smushFile()
    {
        if (!is_file($this->imageLocation) || !is_readable($this->imageLocation)) {
            throw new SmushItException('Could not read image file', 500, $this->imageLocation);
        }

        $jsonStr = $this->client->execute(Client::TYPE_LOCAL, $this->imageLocation);

        return $this->parseResponse($jsonStr);
    }

    /**
     * Parse the response from the Smush.it web service.
     *
     * @param  string $jsonStr A JSON string response from the Smush.it web service.
     *
     * @throws SmushItException
     * @return \stdClass
     */
    private function parseResponse($jsonStr)
    {
        $result = json_decode($jsonStr);

        if (null === $result) {
            throw new SmushItException('Bad response received from the Smush.it service.', 500, $this->imageLocation);
        }

        if (isset($result->error)) {
            throw new SmushItException($result->error_long, $result->error, $this->imageLocation);
        }

        return $result;
    }

}


src/Client/Curl.php
<?php

namespace DavGothic\SmushIt\Client;

use DavGothic\SmushIt\SmushIt;

/**
 * A cURL SmushIt client
 *
 * @codeCoverageIgnore
 */
class Curl extends Client
{
    /**
     * Curl constructor.
     *
     * @throws \RuntimeException If the cURL PHP extension is not loaded.
     */
    public function __construct()
    {
        // @codeCoverageIgnoreStart
        if (!extension_loaded('curl')) {
            throw new \RuntimeException('The cURL PHP extension was not found.');
        }
        // @codeCoverageIgnoreEnd
    }

    /**
     * @inheritdoc
     */
    public function execute($requestType, $imageLocation)
    {
        $curl = curl_init();
        curl_setopt($curl, CURLOPT_RETURNTRANSFER, true);
        curl_setopt($curl, CURLOPT_CONNECTTIMEOUT, $this->timeout);
        curl_setopt($curl, CURLOPT_USERAGENT, $this->userAgent);

        if (self::TYPE_LOCAL === $requestType) {
            curl_setopt($curl, CURLOPT_URL, SmushIt::SMUSH_URL);
            curl_setopt($curl, CURLOPT_POST, true);
            curl_setopt($curl, CURLOPT_POSTFIELDS, ['files' => new \CURLFile($imageLocation)]);
        } else {
            curl_setopt($curl, CURLOPT_URL, SmushIt::SMUSH_URL . http_build_query([
                    'img' => $imageLocation,
                ]));
        }

        return curl_exec($curl);
    }

}


src/Client/Guzzle.php
<?php

namespace DavGothic\SmushIt\Client;

use DavGothic\SmushIt\SmushIt;
use GuzzleHttp\Client as GuzzleClient;

/**
 * A Guzzle SmushIt client
 *
 * @codeCoverageIgnore
 */
class Guzzle extends Client
{

    /**
     * @var GuzzleClient|null
     */
    private $guzzleClient;

    /**
     * Guzzle constructor.
     *
     * @param GuzzleClient|null $guzzleClient A GuzzleClient instance.
     */
    public function __construct(GuzzleClient $guzzleClient = null)
    {
        if (null === $guzzleClient) {
            $this->guzzleClient = new GuzzleClient(['timeout' => $this->timeout]);
        } else {
            $this->guzzleClient = $guzzleClient;
        }
    }

    /**
     * @inheritdoc
     */
    public function execute($requestType, $imageLocation)
    {
        if (self::TYPE_LOCAL === $requestType) {
            $res = $this->guzzleClient->post(SmushIt::SMUSH_URL, [
                'multipart' => [
                    [
                        'name'     => 'files',
                        'contents' => fopen($imageLocation, 'rb'),
                    ],
                ],
            ]);
        } else {
            $res = $this->guzzleClient->get(SmushIt::SMUSH_URL, [
                'query' => [
                    'img' => $imageLocation,
                ],
            ]);
        }

        return $res->getBody();
    }

}


src/Client/ClientInterface.php
<?php

namespace DavGothic\SmushIt\Client;

interface ClientInterface
{

    /**
     * Get the user agent string used for the request.
     *
     * @return string
     */
    public function getUserAgent();

    /**
     * Set the user agent string to use for the request.
     *
     * @param string $userAgent The user agent string.
     */
    public function setUserAgent($userAgent);

    /**
     * Get the request timeout.
     *
     * @return int
     */
    public function getTimeout();

    /**
     * Set the request timeout.
     *
     * @param int $timeout The timeout in seconds.
     */
    public function setTimeout($timeout);

    /**
     * @param int $requestType
     * @param string $imageLocation
     *
     * @return mixed
     */
    public function execute($requestType, $imageLocation);

}


smushit.php
<?php

/**
 * SmushIt - A PHP client for the Yahoo! Smush.it web service.
 *
 * @version    1.3
 * @author     David Hancock <davgothic@gmail.com>
 * @author     Elan Ruusamäe <glen@delfi.ee>
 * @copyright  (c) 2011 David Hancock
 * @license    MIT
 * @link       http://github.com/davgothic/SmushIt
 */
class SmushIt
{

    // URL to the Smush.it web service.
    const SMUSH_URL = 'http://api.resmush.it/ws.php?';

    // User agent string to set for the request.
    const USER_AGENT = 'ShushIt PHP Client/1.3 (+http://github.com/davgothic/SmushIt)';

    /**
     * @var string location of the image.
     */
    private $imageLocation;

    /**
     * @var resource The cURL handle.
     */
    private $curl;

    /**
     * @var int Time of last request.
     */
    private $requestTime;

    /**
     * @var int How often it is allowed to send requests. In microseconds.
     */
    public $requestInterval = 1000000;

    /**
     * Create a new SmushIt instance.
     *
     * @throws \RuntimeException
     */
    public function __construct()
    {
        if ( ! extension_loaded('json')) {
            throw new \RuntimeException('The JSON extension was not found.');
        }

        if ( ! extension_loaded('curl')) {
            throw new \RuntimeException('The cURL extension was not found.');
        }

        $this->curl = curl_init();
        curl_setopt($this->curl, CURLOPT_RETURNTRANSFER, true);
        curl_setopt($this->curl, CURLOPT_CONNECTTIMEOUT, 5);
        curl_setopt($this->curl, CURLOPT_USERAGENT, self::USER_AGENT);
    }

    /**
     * @param string $imageLocation The location of the image.
     *
     * @return object
     *
     *  src       = Source location of the input image.
     *  src_size  = Size of the source image in bytes.
     *  dest      = Temporary location of the compressed image.
     *  dest_size = Size of the compressed image in bytes.
     *  percent   = How much smaller the compressed image is.
     */
    public function compress($imageLocation)
    {
        // Check if we should throttle the request to once per $requestInterval
        if ( ! empty($this->requestTime)) {
            $sinceLast = ((microtime(true) - $this->requestTime) * 1000000);

            if ($sinceLast < $this->requestInterval) {
                // Sleep it off
                usleep($this->requestInterval - $sinceLast);
            }
        }

        $this->requestTime = microtime(true);

        $this->imageLocation = $imageLocation;

        if (preg_match('/https?:\/\//', $this->imageLocation) === 1) {
            $result = $this->smushUrl();
        } else {
            $result = $this->smushFile();
        }

        return $result;
    }

    /**
     * Compress a remote image using the Smush.it web service.
     *
     * @return object
     */
    private function smushUrl()
    {
        curl_setopt($this->curl, CURLOPT_URL, self::SMUSH_URL . http_build_query(array('img' => $this->imageLocation)));
        $jsonStr = curl_exec($this->curl);

        return $this->parseResponse($jsonStr);
    }

    /**
     * Compress a local image using the Smush.it web service.
     *
     * @throws SmushItException
     * @return object
     */
    private function smushFile()
    {
        if ( ! is_file($this->imageLocation) || ! is_readable($this->imageLocation)) {
            throw new SmushItException('Could not read image file', $this->imageLocation);
        }

        curl_setopt($this->curl, CURLOPT_URL, self::SMUSH_URL);
        curl_setopt($this->curl, CURLOPT_POST, true);
        curl_setopt($this->curl, CURLOPT_POSTFIELDS, array('files' => '@' . $this->imageLocation));
        $jsonStr = curl_exec($this->curl);

        return $this->parseResponse($jsonStr);
    }

    /**
     * Parse the response from the Smush.it web service.
     *
     * @param  string $jsonStr A JSON string response from the Smush.it web service.
     *
     * @throws SmushItException
     * @return object
     */
    private function parseResponse($jsonStr)
    {
        $result = json_decode($jsonStr);

        if (is_null($result)) {
            throw new SmushItException('Bad response received from the Smush.it service.', $this->imageLocation);
        }

        if (isset($result->error)) {
            throw new SmushItException($result->error, $this->imageLocation);
        }

        return $result;
    }

}

/**
 * SmushIt exception.
 */
class SmushItException extends \Exception
{

    /**
     * @var string Location of the image.
     */
    private $image;

    /**
     * Creates a new exception.
     *
     * @param string $message Error message.
     * @param string $image   Location of the image.
     */
    public function __construct($message, $image)
    {
        $this->image = $image;
        parent::__construct($message);
    }

    /**
     * Get the location of the image.
     *
     * @return string
     */
    final public function getImage()
    {
        return $this->image;
    }

}
